tests/test_template.py

```diff
@@ -0,0 +1,135 @@
+        return np.random.randn(*arg) + 1j*np.random.randn(*arg)
+
+    def setUp(self):
+        super().setUp()
+        np.random.seed(35)
+        self.u1 = self.random_interferometer(self.num_subsystems)
+        self.u2 = self.random_interferometer(self.num_subsystems-1)
+
+    def test_interferometer(self):
+        dev = qml.device('default.gaussian', wires=self.num_subsystems)
+        np.random.seed(8)
+        squeezings = np.random.rand(self.num_subsystems, 2)
+        num_params = int(self.num_subsystems*(self.num_subsystems-1)/2)
+        theta = np.random.uniform(0, 2*np.pi, num_params)
+        phi = np.random.uniform(0, 2*np.pi, num_params)
+
+        @qml.qnode(dev)
+        def circuit(theta, phi):
+            for wire in range(self.num_subsystems):
+                qml.Squeezing(squeezings[wire][0], squeezings[wire][1], wires=wire)
+
+            qml.template.Interferometer(theta=theta, phi=phi, wires=range(self.num_subsystems))
+            return tuple(qml.expval.MeanPhoton(wires=wires) for wires in range(self.num_subsystems))
+
+        self.assertAllAlmostEqual(circuit(theta, phi), [0.96852694, 0.23878521, 0.82310606, 0.16547786], delta=self.tol)
+
+        self.assertAllAlmostEqual(qml.jacobian(circuit, 0)(theta, phi),
+                                  np.array([[-6.18547621e-03, -3.20488416e-04, -4.20274088e-02, -6.21819677e-02, 0.00000000e+00,  0.00000000e+00],
+                                            [ 3.55439469e-04,  3.89820231e-02, -3.35281297e-03,  7.93009278e-04, 8.30347900e-02, -3.45150712e-01],
+                                            [ 5.44893709e-03,  9.30878023e-03, -5.33374090e-01,  6.13889584e-02,-1.16931386e-01,  3.45150712e-01],
+                                            [ 3.81099656e-04, -4.79703149e-02,  5.78754312e-01,  0.00000000e+00, 3.38965962e-02,  0.00000000e+00]]
+                                  ), delta=self.tol)
+
+    def setUp(self):
+        super().setUp()
+        np.random.seed(8)
+        self.num_wires = 4
+
+    def test_cvqnn(self):
+
+        dev = qml.device('default.gaussian', wires=self.num_wires)
+
+        weights = [[
+            np.random.uniform(0, 2*np.pi, int(self.num_wires*(self.num_wires-1)/2)), #transmittivity angles
+            np.random.uniform(0, 2*np.pi, int(self.num_wires*(self.num_wires-1)/2)), #phase angles
+            np.random.uniform(0.1, 0.7, self.num_wires), #squeezing amounts
+            np.random.uniform(0, 2*np.pi, int(self.num_wires*(self.num_wires-1)/2)), #transmittivity angles
+            np.random.uniform(0, 2*np.pi, int(self.num_wires*(self.num_wires-1)/2)), #phase angles
+            np.random.uniform(0.1, 2.0, self.num_wires), #displacement magnitudes
+            np.random.uniform(0.1, 0.3, self.num_wires) #kerr parameters
+        ] for l in range(2)]
+
+        with patch.object(Kerr, '__init__', return_value=None) as _: #Kerr() does not work on any core device, so we have to mock it here with a trivial class
```

Comment:
I'm a bit uncomfortable with this approach, but I can't think of a better way to test :thinking: 

Also, we should add a test of a non-mocked QNN to the pennylane-sf repo

Reply:
I think I agree. The only possible objection could be: Would this not me more a test of the qnode, rather than the template? Also I this should be done for all test and thus maybe better in a separate PR and only after 0.2?

Reply:
Yes, I suppose it still wouldn't be a true unit test, but more of an integration test between the template function and the QNode. I can't think of a better way to unit test the template function though...